### ISSUE TITLE
decide on text states storage requirements

### ISSUE THREAD

**Issue by irina060981** [alpheios-project/alignment-editor-new#2]
@balmas and @kirlat , I suggest to discuss overall scope of the task here.

**About handling with texts:**

After examinining the requirements I believe that we could define the following states of source/translations text:

1. input text 
    - it could be uploaded manually, from file, from external API (DTS API)
    - it could be a plain text, TEI XML or text with already defined alignment based on Alpheios Alignment Schema
    - it could have metadata in xml format (TEI XML)

2.  updated text after upload 
    - a user could edit line breaks
    - could add some words
    - metada could be editted

3.  tokenized text before alignment ( it could be fully tokenized or partly according to the discussion in #1 )
    - it should be able to edit tokenized results
    - add line breaks
    - add words

4.  text on each alignment iteration
    - a user should be able to undo/redo each alignment step 
    - should be able to add comments to aligned groups
    
5.  output alignment result (with various additional features)

and also we could have several translations texts for one source texts

To realize such requirements we could go one of the following ways:

- we could store texts on each state - so we would have 3 versions + all alignment steps
    **advantages** - it would be easier to handle, manupulate and go between steps
    **disadvantages** - it would use much memory

- we could store transactions of texts changing
      for example, 

            - plain text
            - added line break
            - added words
            - added token to this word
            - added new words
            - added another token
            - added alignment for these tokens
     and when we need to to get the text on any state we apply all these changes
     **advantages** - for big texts it would need less memory
     **disadvantages** - it would need additional calculations

I like the first variant more, but I think that we could use some storage system in addition to it
    - relational DB or IndexedDB

At some point we could try to store it in memory in some array/object-like structures

Or we could define what states could not be re-get, for example we store only finally editted input (with/or without tokens) and alignment steps

@balmas, upon requirements what do you think - what amount of stored data in memory we need?
and what states of texts we should store for logged in users?

**Comment by kirlat** [comment]
> 1. input text
> ... 
>   * it could have metadata in xml format (TEI XML)

Can there be any other metadata formats? I think it would be important to clearly define what formats (or their subsets) we support and what not. This can be especially significant as we probably will need to create sets of rules to verify validity of input documents: they can come from various sources and we should be ready to expect anything.

This will also help us to understand better the subset of data we'll be working with. With this, we can decide how to store and process data more efficiently.

**Comment by kirlat** [comment]
How would we store data in memory? Would it be a JSON-like structure? We could probably store data in JSON-like representation by making each word a node and having `before` and `after` props that will keep XML/HTML/etc. tags that are placed before and after the word respectively. It will allow us to keep text in a plain-text-like format (i.e. have words only) and yet retain all formatting and metadata. Of course, it will require some memory to be available.

I'm maybe delving into details too early, but if we represent our data in JSON, we could use any of the existing finite state machine implementations to keep the history of the state of data. Solutions include, but are not limited to, [automerge](https://github.com/automerge/automerge) (automerge is mostly for merging two states, which can be helpful in some situations too, but it can also keep history and provide the undo functionality) or [Redux](https://redux.js.org/). I don't think we'll be able to write something within a reasonable amount of time that will be able to fulfill the task better. 

Using a finite state machine we won't have to chose what to keep: full state snapshots or incremental transactions. That all will be handled inside the machine itself. We'll have access to both the transactions and will also be able to obtain any previous state that was recorded. Maybe that would be a sensible approach? What do you think?

**Comment by irina060981** [comment]
>Can there be any other metadata formats? I think it would be important to clearly define what formats (or their subsets) we support and what not. This can be especially significant as we probably will need to create sets of rules to verify validity of input documents: they can come from various sources and we should be ready to expect anything.
This will also help us to understand better the subset of data we'll be working with. With this, we can decide how to store and process data more efficiently.

According to the current requirements (https://github.com/alpheios-project/documentation/blob/master/development/alignment-editor.md):

```
Must allow user to add identifying bibliographic metadata about source and translation texts (e.g. author, title, edition, etc.)

May support automatic detection of metadata provided in Dublin Core or other standard in the source input text (in the case of TEI input)
```

I think it is not possible to define all formats forever.
I believe it would depend on real users and there texts, so I think it would be better to create a posibility to define standard on upload and transform according to this

**Comment by irina060981** [comment]
>How would we store data in memory? Would it be a JSON-like structure? We could probably store data in JSON-like representation by making each word a node and having before and after props that will keep XML/HTML/etc. tags that are placed before and after the word respectively. It will allow us to keep text in a plain-text-like format (i.e. have words only) and yet retain all formatting and metadata. Of course, it will require some memory to be available.

I think that it is worth to decide it in requirements first to avoid unuseful complexity

**Comment by balmas** [comment]
> > Can there be any other metadata formats? I think it would be important to clearly define what formats (or their subsets) we support and what not. This can be especially significant as we probably will need to create sets of rules to verify validity of input documents: they can come from various sources and we should be ready to expect anything.
> > This will also help us to understand better the subset of data we'll be working with. With this, we can decide how to store and process data more efficiently.
> 
> According to the current requirements (https://github.com/alpheios-project/documentation/blob/master/development/alignment-editor.md):
> 
> ```
> Must allow user to add identifying bibliographic metadata about source and translation texts (e.g. author, title, edition, etc.)
> 
> May support automatic detection of metadata provided in Dublin Core or other standard in the source input text (in the case of TEI input)
> ```
> 
> I think it is not possible to define all formats forever.
> I believe it would depend on real users and there texts, so I think it would be better to create a posibility to define standard on upload and transform according to this

We should support metadata provided in at least the following ways:

* within elements in the teiHeader of the TEI XML document 
* for texts retrieved from a DTS API, in the dts:dublincore and dts:extensions properties from a [DTS API collections endoint](https://distributed-text-services.github.io/specifications/Collections-Endpoint.html)
* as user entered name/value pairs

In the first two cases, we need to allow the user to edit any metadata that was automatically extracted from the XML or DTS API as it's unlikely we will be able to get this 100% correct.  Internally, it can probably be modeled as key/value store. But note that we will want to support full URIs for metadata properties,  and probably should offer users the option to select terms from the Dublin Core controlled vocabulary (https://www.dublincore.org/specifications/dublin-core/dcmi-terms/)

**Comment by balmas** [comment]
Notes from today's discussion:

1. We would like to have the option to retain the original text as provided by the user (whether in TEI XML, plain text, or an alignment XML) as the source of an alignment. This option might require login (or eventually a higher level of service) but we should make sure our design can accommodate this even if it's not offered by default.

2. When the words are extracted from the source text, our data model should retain a link between the words and their source.  I would like us to consider whether we can use the W3C annotation selector models for this, as per the diagram at https://github.com/alpheios-project/alignment-editor-new/issues/1#issuecomment-655016942 . I think it would be out of scope to try to update those links if the words are edited after the initial extraction from the source text, but by retaining the initial link we can at least provide a means to get back to the original text.

3. Once words are extracted in our alignment model, we should have the ability to track state changes with our own data model. We need to talk more about how this is done, and to what extent they are retained beyond the browser session.

**Comment by balmas** [comment]
Expanding upon the conversation of where to store state and data:

Options for storage:
| Where | Limitations | Candidates |
| --- | --- |---|
| Browser memory | lasts only as long as the page is loaded, available size depends upon local memory, large amounts of data will affect performance | current alignment state, current alignment data (full or partial)|
| Browser local indexdb | lasts until user clears their cache, size depends upon local disk | current alignment state, current alignment data (full or partial), prior alignments |
| Local file system | user can delete at any time, size depends upon local disk, browser access limited | prior alignments |
| Remote Database | fully under our control, available size depends upon what we want to support, requires a network connection, network speed, data size affects performance | current alignment state, current alignment data (full), prior alignments |

**Comment by balmas** [comment]
(Note I have moved the metadata requirements discussion to a new issue #4 )

**Comment by balmas** [comment]
assigning to myself to decide if we want to support a local indexed db option or if it will be only memory vs remote

**Comment by balmas** [comment]
per discussion at today's check-in
* we are going to need indexeddb to support in-progress work 
* local indexdb interface does not have to (and probably should not) mirror remote storage interface
* need to think more about shared workstation scenarios

**Comment by balmas** [comment]
See requirements in #19

**Comment by irina060981** [comment]
Comments/Questions are published inside #19 
@balmas , should I answer here anything?

**Comment by balmas** [comment]
No I think the requirements are pretty much set.   I'll close this and we can work through the implementation via #19 and subsequent.
